src/core/field_coupling/fields/Interpolated.hpp

```diff
@@ -0,0 +1,145 @@
+#ifndef CORE_EXTERNAL_FIELD_FIELDS_INTERPOLATED_HPP
+#define CORE_EXTERNAL_FIELD_FIELDS_INTERPOLATED_HPP
+
+#include "gradient_type.hpp"
+
+#include "Vector.hpp"
+
+#include "utils/interpolation/bspline_3d.hpp"
+#include "utils/interpolation/bspline_3d_gradient.hpp"
+#include "utils/math/tensor_product.hpp"
+
+/* Turn off range checks if release build. */
+#if defined(NDEBUG) && !defined(BOOST_DISABLE_ASSERTS)
+#define BOOST_DISABLE_ASSERTS
+#endif
+#include <boost/multi_array.hpp>
+
+#include <array>
+
+namespace FieldCoupling {
+namespace Fields {
+namespace detail {
+template <typename T>
+void deep_copy(boost::multi_array<T, 3> &dst,
+               const boost::multi_array<T, 3> &src) {
+  auto *s = src.shape();
+  dst.resize(boost::extents[s[0]][s[1]][s[2]]);
+  dst = src;
+
+  auto *b = src.index_bases();
+  dst.reindex(std::array<typename boost::multi_array<T, 3>::index, 3>{
+      b[0], b[1], b[2]});
+}
+
+template <typename F, class = void>
+struct is_linear : public std::false_type {};
+
+template <typename F>
+struct is_linear<F, typename std::enable_if<F::is_linear>::type>
+    : std::true_type {};
+}
+
+/**
+* @brief A vector field interpolated from a regular grid.
+*
+* This is an interpolation wrapper around a boost::multi_array,
+* which can be evaluated on any point in space by spline interpolation.
+*/
+template <typename T, size_t codim> class Interpolated {
+public:
+  using value_type = typename decay_to_scalar<Vector<codim, T>>::type;
+  using gradient_type = detail::gradient_type<T, codim>;
+  using storage_type = boost::multi_array<value_type, 3>;
```

Comment:
Some time ago I've written a `constexpr`-friendly statically sized multidimensional array.  In other words, an N-dimensional `std::array`.  It looks like here everything is 3D, so `boost::multi_array` might incur an additional overhead for runtime allocation.
My code is also usable in CUDA.  License is CC0:
https://github.com/hmenke/cpp-patterns/blob/master/multi_array/multi_array.hpp
This code will find its way into Espresso sooner or later because I make extensive use of it in my SD implementation.

Reply:
Actually the dim is static in `multi_array`, so there is not cost. It's terrible for a different reason: it has strange copy semantics and is not movable. So we'll replace this sooner or later.